src/dp.rs

```diff
@@ -1,16 +1,57 @@
 // SPDX-License-Identifier: MPL-2.0
 
 //! Differential privacy (DP) primitives.
-use std::fmt::Debug;
+use num_bigint::{BigInt, BigUint, TryFromBigIntError};
+use num_rational::{BigRational, Ratio};
 
-/// Positive rational number to represent DP and noise distribution parameters in protocol messages
-/// and manipulate them without rounding errors.
+/// Errors propagated by methods in this module.
+#[derive(Debug, thiserror::Error)]
+pub enum DpError {
+    /// Tried to use an infinite float as privacy parameter.
+    #[error("DP error: input float was infinite.")]
+    InfinityFloat(),
+
+    /// Tried to construct a rational number with zero denominator.
+    #[error("DP error: input denominator was zero.")]
+    ZeroDenominator(),
+
+    /// Tried to convert BigInt into something incompatible.
+    #[error("DP error: {0}")]
+    BigIntConversion(#[from] TryFromBigIntError<BigInt>),
+}
+
+/// Positive finite precision rational number to represent DP and noise distribution parameters in
+/// protocol messages and manipulate them without rounding errors.
 #[derive(Clone, Debug)]
-pub struct Rational {
-    /// Numerator.
-    pub numerator: u32,
-    /// Denominator.
-    pub denominator: u32,
+pub struct Rational(Ratio<BigUint>);
+
+impl Rational {
```

Comment:
Maybe `URational`? Also, I see many functions that start by verifying that the denominator is not 0. It's good to be safe but I was wondering if this can even happen if we never use `new_raw` (the default constructor and `recip` will panic if we divide by 0: https://docs.rs/num-rational/latest/num_rational/struct.Ratio.html#method.new)

Reply:
i think the check in the `DiscreteGaussian` constructor is nice because it prevents a panic in the sampler. the checks inside the sampler are just a way to enforce what is said in the docstrings and not expected to trigger when the API is used. is it customary for constructors to panic instead of error if you give them something invalid?

Reply:
Yes, we should definitely check in the `DiscreteGaussian` constructor and return an error. My question was more about the checks inside the sampler, since they are not supposed to be triggered when the API is used. But it's nice to show that the docstrings are properly enforced, so why not!

Reply:
@tholop -- just trying to make sure comments are getting resolved.